Add a DatePicker component

ismailElkalili opened the pull request.

ismailElkalili committed:
Add a DatePicker component using Material Design 3 that includes Clear, OK, and Cancel buttons.
- The DatePicker should initialize with the default (current) date selected.
- Implement custom colors for the selected date in the picker.

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/designSystem/component/DatePicker.kt:
```diff
@@ -0,0 +1,115 @@
+                    }
+
+                    Spacer(modifier = Modifier.weight(1f))
+
+                    TextButton(onClick = { showDialog = false }) {
+                        Text(text = "Cancel",
+                            style =  TudeeTheme.testStyle.label.large,
+                            color = TudeeTheme.color.primary)
+                    }
+
+                    TextButton(onClick = {
+                        val millis = datePickerState.selectedDateMillis
+                        if (millis != null) {
+                            selectedDate = LocalDate.ofEpochDay(millis / (24 * 60 * 60 * 1000))
+                        }
+                        showDialog = false
+                    }) {
+                        Text(text = "OK",
+                            style =  TudeeTheme.testStyle.label.large,
+                            color = TudeeTheme.color.primary)
+                    }
+                }
```
The button texts "`Clear`", "`Cancel`", and "`OK`" should come from string resources to support localization and consistency.

ismailElkalili replied:
I didn't implement it because we haven't set up a strings.xml file yet, and doing so at this stage could create a lot of conflicts

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/designSystem/component/DatePicker.kt:
```diff
@@ -0,0 +1,115 @@
+                        todayDateBorderColor = TudeeTheme.color.primary,
+                    ),)
+
+                Spacer(modifier = Modifier.height(8.dp))
```
Try to avoid using extra composables like `Spacer` when a simple padding on the surrounding composable's `Modifier` can achieve the same result

ismailElkalili replied:
also it's the same situation here. I noticed that some developers have their own custom Spacer implementations, so if I add mine now, it could lead to conflicts. I'm trying to avoid that for the time being until we finalize the project structure and properly organize everything

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/designSystem/component/DatePicker.kt:
```diff
@@ -0,0 +1,115 @@
+@OptIn(ExperimentalMaterial3Api::class)
+fun DatePicker() {
+    var selectedDate by remember { mutableStateOf<LocalDate?>(null) }
+    var showDialog by remember { mutableStateOf(false) }
```
Since we're following the MVVM pattern, it's better to pass the showDialog state from outside the `composable`. The view model should handle the `interaction `— for example, when a button is clicked, the interaction triggers a function like `onClickShowDatePicker`() in the view model, which `updates `the `showDatePickerDialog `state. The composable should simply observe that state and react by showing the dialog when needed. This keeps the UI stateless and properly tied to the app’s state management flow.

ismailElkalili replied:
u are absolutely right , this logic should be handled in the ViewModel to keep the composable stateless and fully aligned with MVVM.
but for now we make component so I kept it inside the component for simplicity during early development, but we refactor it to move the state and interaction logic into the ViewModel when make ViewModel

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/designSystem/component/DatePicker.kt:
```diff
@@ -0,0 +1,115 @@
+                .atStartOfDay(ZoneId.systemDefault()).toInstant().toEpochMilli()
+    )
+
+    if (showDialog) {
```
Instead of using a simple if (showDialog), it might be better to use AnimatedVisibility. That way, you can add smooth transitions when the dialog appears or disappears, which can improve the overall user experience.

ismailElkalili replied:
I search for AnimatedVisibility with DatePickerDialog like i uesed and i found i can't uesing AnimatedVisibility with DatePickerDialog but i can use with DatePicker

Belal-Alkhatib replied:
You can actually use `AnimatedVisibility` inside any `composable`, including where you're showing the `DatePickerDialog`. Just replace the` if (showDialog)` block with `AnimatedVisibility(visible = showDialog)`. If you need custom animation behavior, you can tweak the enter and exit transition parameters. 

This makes the visible part feel more alive, not just something that suddenly appears

ismailElkalili replied:
u are right in general, but not when it comes to popups. i am using DatePickerDialog which is shown as a popup soooo the animation doesn't work the way we expect since it is rendered outside the normal screen layout. sooo i uesed  if (showDialog)
and u can search about that if u don't convinced by the answer hahaha

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/designSystem/component/DatePicker.kt:
```diff
@@ -0,0 +1,115 @@
+
+    if (showDialog) {
+        DatePickerDialog(
+            onDismissRequest = { showDialog = false },
```
Also here, it’s better not to update `showDialog` directly inside the composable. Instead, delegate the dismissal action to a callback (e.g., `onDismiss`) and let the `view model` handle the state update. This keeps the composable stateless and aligned with unidirectional data flow.

ismailElkalili replied:
The same comment applies regarding the ViewModel that I mentioned above,
so when we implement the ViewModel, we will refactor accordingly, but we are still working within the component

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/designSystem/component/DatePicker.kt:
```diff
@@ -0,0 +1,115 @@
+                Spacer(modifier = Modifier.height(8.dp))
+
+                Row {
+                    TextButton(onClick = {
+                        selectedDate = null
+                        showDialog = false
+                    }) {
+                        Text(text = "Clear",
+                            style =  TudeeTheme.testStyle.label.large,
+                            color = TudeeTheme.color.primary)
+                    }
```
avoid setting selectedDate and showDialog directly inside the composable. Instead, trigger a callback like `onClearClick`()

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/designSystem/component/DatePicker.kt:
```diff
@@ -0,0 +1,115 @@
+                Spacer(modifier = Modifier.height(8.dp))
+
+                Row {
+                    TextButton(onClick = {
+                        selectedDate = null
+                        showDialog = false
+                    }) {
```
You can replace `TextButton `with a `Text `using `Modifier.clickable` and styling for slightly better performance — especially since we’re using a custom design system and don’t need Material defaults. **Avoiding unnecessary composables helps improve UI efficiency.**

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/designSystem/component/DatePicker.kt:
```diff
@@ -0,0 +1,115 @@
+    }
+}
+
+object TudeeTheme {
+    val color: TudeeColors
+        @Composable
+        get() = LocalTudeeColors.current
+
+    val testStyle: TudeeTextStyle
+        @Composable
+        get() = LocalTudeeTextStyle.current
+}
+
+
+
```
Why is the TudeeTheme object defined here in the DatePicker file?

ismailElkalili committed:
Refactor PR commit:
- replace TextButton with Text using Modifier.clickable
- replace local TudeeTheme object to Theme reference
- replace Spacer to Modifier padding

ismailElkalili committed:
Refactor PR commit:
Removed internal state

Belal-Alkhatib commented on app/src/main/java/com/qudus/tudee/designSystem/component/DatePicker.kt:
```diff
@@ -0,0 +1,117 @@
+                            horizontalArrangement = Arrangement.SpaceBetween,
+                    verticalAlignment = Alignment.CenterVertically
+                ) {
+                    Text(
+                    text = "Clear",
+                        modifier = Modifier.clickable {
+                            onClear()
+                            onDismiss()
+                    },
+                        style =  Theme.textStyle.label.large,
+                        color = Theme.color.primary)
+                    }
+
+                    Spacer(modifier = Modifier.weight(1f))
+
+                Row() {
+                    Text(text = "Cancel",
+                        style =  Theme.textStyle.label.large,
+                        color = Theme.color.primary,
+                        modifier = Modifier.clickable { onDismiss() })
+
+
+                Text(
+                    text = "OK",
+                    modifier = Modifier.clickable {
+                        val millis = datePickerState.selectedDateMillis
+                        if (millis != null) {
+                            val date = LocalDate.ofEpochDay(millis / (24 * 60 * 60 * 1000))
+                            onDateSelected(date)
+                        }
+                        onDismiss()
+                    },
+                    style = Theme.textStyle.label.large,
+                    color =Theme.color.primary
+                )
+                }
+                }
```
Again, using hardcoded text like "`Clear`", "`Cancel`", and "`OK`" makes it difficult to support multiple languages. It’s better to use string resources (`stringResource(R.string.clear), etc`.) to make the app easier to localize and maintain.

Belal-Alkhatib reviewed (changes requested):
good work ismail
